onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlSTFT.h

```diff
@@ -0,0 +1,610 @@
+
+        // input 0: signal (required; tensor)
+        {
+            // Signal shape is expected to be [batch_size, signal_length, 1] or [batch_size, signal_length] for
+            // real-valued input. It must be [batch_size, signal_length, 2] for complex input.
+            uint32_t rank;
+            ORT_THROW_IF_FAILED(shapeInterface->GetInputTensorDimensionCount(DmlSTFTKernelInputIndex::Signal, &rank));
+            ML_CHECK_VALID_ARGUMENT(rank == 2 || rank == 3, "Signal shape must be 2D or 3D.");
+
+            std::array<uint32_t, 3> dims;
+            ORT_THROW_IF_FAILED(shapeInterface->GetInputTensorShape(DmlSTFTKernelInputIndex::Signal, rank, dims.data()));
+            this->batchSize = dims[0];
+            this->signalSize = dims[1];
+
+            if (rank == 3)
+            {
+                ML_CHECK_VALID_ARGUMENT(dims[2] == 1, "DML STFT only accepts real-valued input.");
+            }
+
+            MLOperatorEdgeDescription edgeDesc;
+            ORT_THROW_IF_FAILED(context->GetInputEdgeDescription(0, &edgeDesc));
+
+            assert(edgeDesc.edgeType == MLOperatorEdgeType::Tensor);
+            this->dataType = Dml::GetDmlDataTypeFromMlDataType(edgeDesc.tensorDataType);
+        }
+
+        // input 1: frame_step (required; constant; scalar)
+        {
+            ComPtr<IMLOperatorTensor> tensorInterface;
+            ORT_THROW_IF_FAILED(context->GetConstantInputTensor(DmlSTFTKernelInputIndex::FrameStep, &tensorInterface));
+
+            MLOperatorTensor tensor(tensorInterface.Get());
+            this->frameStep = onnxruntime::narrow<uint32_t>(OperatorHelper::ReadScalarTensorCastToInt64(tensor));
+        }
+
+        // input 2: window (optional; tensor)
+        if (context->IsInputValid(DmlSTFTKernelInputIndex::Window))
+        {
+            uint32_t rank;
+            ORT_THROW_IF_FAILED(shapeInterface->GetInputTensorDimensionCount(DmlSTFTKernelInputIndex::Window, &rank));
+            ML_CHECK_VALID_ARGUMENT(rank == 1, "Window shape must be 1D.");
+
+            ORT_THROW_IF_FAILED(shapeInterface->GetInputTensorShape(DmlSTFTKernelInputIndex::Window, rank, &this->frameSize));
+
+            this->hasWindowTensor = true;
+        }
+
+        // input 3: frame_length (optional; constant; scalar)
+        if (context->IsInputValid(DmlSTFTKernelInputIndex::FrameLength))
+        {
+            ComPtr<IMLOperatorTensor> tensorInterface;
+            ORT_THROW_IF_FAILED(context->GetConstantInputTensor(DmlSTFTKernelInputIndex::FrameLength, &tensorInterface));
+
+            MLOperatorTensor tensor(tensorInterface.Get());
+            uint32_t frameLength = onnxruntime::narrow<uint32_t>(OperatorHelper::ReadScalarTensorCastToInt64(tensor));
+
+            ML_CHECK_VALID_ARGUMENT(
+                this->frameSize == 0 || this->frameSize == frameLength,
+                "The window size and frame_length must be equal, if both are provided."
+            );
+
+            this->frameSize = frameLength;
```

Comment:
>this->frameSize = frameLength; [](http://example.com/codeflow?start=12&length=30)

I see this overrides line 87's value. It's an error if neither of them are set. Is it also an error if *both* window and frameLength are set?

Reply:
It's OK if they're both set so long as they are the same value (see check above on lines 101-104).